internal/services/gcp_reservation_service.go

```diff
@@ -0,0 +1,139 @@
+	logger := ctxval.Logger(r.Context())
+
+	var accountId int64 = ctxval.AccountId(r.Context())
+
+	// TODO: meanwhile these values are hardcoded until we will have instance types, sources, ssh endpoints,
+	payload := &payloads.GCPReservationRequestPayload{
+		PubkeyID: 2,
+		// TODO: The project id change to source id from sources
+		SourceID:    "citric-expanse-361512",
+		Zone:        "us-central1-a",
+		MachineType: "n1-standard-1",
+		ImageID:     "bc97177c-9d07-4db9-ad59-8b2c0bf4174e",
+		PowerOff:    true,
+		Amount:      2,
+	}
+	// TODO: Uncomment after removing hard-coded values
+	// if err := render.Bind(r, payload); err != nil {
+	// 	renderError(w, r, payloads.NewInvalidRequestError(r.Context(), err))
+	// 	return
+	// }
+
+	rDao, err := dao.GetReservationDao(r.Context())
+	if err != nil {
+		renderError(w, r, payloads.NewInitializeDAOError(r.Context(), "reservation DAO", err))
+		return
+	}
+
+	pkDao, err := dao.GetPubkeyDao(r.Context())
+	if err != nil {
+		renderError(w, r, payloads.NewInitializeDAOError(r.Context(), "pubkey DAO", err))
+		return
+	}
+
+	detail := &models.GCPDetail{
+		Zone:        payload.Zone,
+		MachineType: payload.MachineType,
+		Amount:      payload.Amount,
+		PowerOff:    payload.PowerOff,
+	}
+	reservation := &models.GCPReservation{
+		PubkeyID: payload.PubkeyID,
+		// TODO: The project id change to source id from sources
+		ImageID:  payload.ImageID,
+		SourceID: payload.SourceID,
+		Detail:   detail,
+	}
+
+	reservation.AccountID = accountId
+	reservation.Status = "Created"
+	reservation.Provider = models.ProviderTypeGCP
+	reservation.Steps = 1
+
+	logger.Debug().Msgf("Validating existence of pubkey %d for this account", reservation.PubkeyID)
+	pk, err := pkDao.GetById(r.Context(), reservation.PubkeyID)
+	if err != nil {
+		var e dao.NoRowsError
+		if errors.As(err, &e) {
+			renderError(w, r, payloads.NewNotFoundError(r.Context(), err))
+		} else {
+			renderError(w, r, payloads.NewDAOError(r.Context(), "get pubkey by id", err))
+		}
+		return
+	}
+	logger.Debug().Msgf("Found pubkey %d named '%s'", pk.ID, pk.Name)
+
+	// create reservation in the database
+	err = rDao.CreateGCP(r.Context(), reservation)
+	if err != nil {
+		renderError(w, r, payloads.NewDAOError(r.Context(), "create reservation", err))
+		return
+	}
+	logger.Debug().Msgf("Created a new reservation %d", reservation.ID)
+
+	// TODO: Fetch project id from sources
+	// TODO: upload key job if needed
+
+	// Get Image builder client
+	IBClient, ibErr := clients.GetImageBuilderClient(r.Context())
+	logger.Trace().Msg("Creating IB client")
+	if ibErr != nil {
+		renderError(w, r, payloads.NewClientInitializationError(r.Context(), "image builder client", ibErr))
+		return
+	}
+
+	// Get Image Name
+	name, ibErr := IBClient.GetGCPImageName(r.Context(), reservation.ImageID)
+	if ibErr != nil {
+		renderError(w, r, payloads.ClientError(r.Context(), "Image Builder", "can't get name from image builder", ibErr, 500))
+	}
```

Comment:
This is missing return statement, it returns invalid reply in case of error:

```
{
  "msg": "Image Builder error: can't get name from image builder: fetch image status call: image compose not found: backend service returned not found (404) or no data",
  "request_id": "cchit927qo0jv3l0r19g"
}
{
  "reservation_id": 2,
  "pubkey_id": 1,
  "source_id": "redhat-360508",
  "zone": "us-central1-a",
  "machine_type": "n1-standard-1",
  "amount": 2,
  "image_id": "bc97177c-9d07-4db9-ad59-8b2c0bf4174e",
  "poweroff": true
}
```